src/cryptonote_core/service_node_list.cpp

```diff
@@ -291,6 +291,104 @@ static uint64_t min_recent_height(cryptonote::network_type nettype, uint64_t hei
     return result;
 }
 
+// NOTE: In debug mode, check that the total locked SESH as stored in the SQL rewards DB matches
+// the amount of locked SESH that is currently recorded on the SNL. We only verify if the SNL
+// height matches the rewards DB height (e.g. the 2 are in sync).
+//
+// The values we are verifying are do not contribute to chain consensus, they are auxiliary values
+// like the amount of locked SESH, the amount of timelocked SESH for analytics purposes hence gating
+// these checks into debug builds.
+static void verify_rewards_db_values(
+        [[maybe_unused]] cryptonote::Blockchain& blockchain,
+        [[maybe_unused]] const service_node_list::state_t& state) {
+#if !defined(NDEBUG)
+    cryptonote::BlockchainSQLite& db = blockchain.sqlite_db();
+    if (blockchain.get_network_version() >= hf::hf21_eth && db.height == state.height) {
+        struct pubkey_stake {
+            crypto::ed25519_public_key pubkey;  // The SN pubkey
+            uint64_t stake;                     // Amount staked for SN pubkey for the ETH address
+        };
+
+        struct stake_history {
+            uint64_t total;                    // Total SESH the ETH address staked
+            std::vector<pubkey_stake> stakes;  // Individual stakes that the ETH address contributed
+        };
+
+        // NOTE: Enumerate all the current SESH that an ETH address has locked up in the network
+        std::unordered_map<eth::address, stake_history> eth_to_locked_stakes;
+
+        // NOTE: Enumerate SNL
+        for (auto it : state.service_nodes_infos) {
+            for (auto contrib_it : it.second->contributors) {
```

Comment:
```suggestion
        for (const auto& [pk, sni] : state.service_nodes_infos) {
            for (const auto& contributor : sni->contributors) {
```

Reply:
I think the misunderstanding here (and in the other cases) is that `auto` has to be opt-in on capturing a reference and so these loop iterations are copying:

```c++
#include <iostream>
#include <vector>

struct A {
    A() { std::cout << "A ctor\n"; }
    A(A&&) { std::cout << "A move ctor\n"; }
    A(const A&) { std::cout << "A copy ctor\n"; }
    A& operator=(const A&) { std::cout << "A copy assign\n"; return *this; }
    A& operator=(A&&) { std::cout << "A move assign\n"; return *this; }
};

int main() {
    std::vector<A> v;
    v.resize(10);
    for (auto x : v) {}
}
```
```
$ ./a.out
A ctor
A ctor
A ctor
A copy ctor
A copy ctor
A copy ctor
```
but with `for (auto& x : v) {}` does not:
```
A ctor
A ctor
A ctor
```